### PR TITLE
std.fmt: fix formatting slices of structs with custom formatting

### PR DESCRIPTION
**Pull request by chadwain**
Fixes `hasMethod` which was introduced in #18636 
It should return false for pointers to more than one item.

### CODE REVIEW

**Comment by Rexicon226** [comment]
I can't comment on the correctness of the fix, but please add a test case :)

**Comment by nektro** [comment]
looks good with a test, since there's only a few options can you change `else` to `.Many, .Slice, .C` when u push the update?

**Comment by chadwain** [comment]
I added test cases for hasFn too since there were none

File: lib/std/meta.zig [hunk 1]
```diff
@@ -1129,15 +1129,78 @@ pub inline fn hasFn(comptime T: type, comptime name: []const u8) bool {
...
+    try std.testing.expect(!hasMethod(u32, "foo"));
+    try std.testing.expect(!hasMethod([]u32, "len"));
+    try std.testing.expect(!hasMethod(struct { u32, u64 }, "len"));
+
+    const S1 = struct {
+        pub fn foo() void {}
+    };
+
+    try std.testing.expect(hasMethod(S1, "foo"));
+    try std.testing.expect(hasMethod(*S1, "foo"));
```
Comment (nektro) [inline lib/std/meta.zig:1176]: oh this function is much more flawed than i first realized, though it is the inner part of the actual correct algorithm.

the first should only pass if `foo` was `pub fn foo(_: S1) void {}` or `pub fn foo(_: S2) void {}`
the second should never pass.

**Review by nektro — CHANGES_REQUESTED** [review]

File: lib/std/meta.zig (hunk 1 above)
Comment (nektro) [reply lib/std/meta.zig:1176]: pretty sure this'll do it

```zig
pub inline fn hasMethod(comptime T: type, comptime name: []const u8) ?bool {
    return comptime switch (@typeInfo(T)) {
        .Struct, .Union, .Enum => blk: {
            if (!hasFn(T, name)) break :blk false;
            const params = @typeInfo(@TypeOf(@field(T, name))).Fn.params;
            if (params.len == 0) break :blk false;
            const First = params[0].type orelse break :blk null;
            break :blk switch (@typeInfo(First)) {
                .Pointer => |P| switch (P.size) {
                    .One => hasFn(P.child, name),
                    .Many, .Slice, .C => false,
                },
                .Struct, .Enum, .Union => First == T,
                else => false,
            };
        },
        else => false,
    };
}
```

Comment (chadwain) [reply lib/std/meta.zig:1176]: If I'm not mistaken, that second test case with `*S1` was the main reason the `hasMethod` function was introduced.

Comment (chadwain) [reply lib/std/meta.zig:1176]: Would it be okay to keep the function as it is, but just give it a better name? (something like `canCallFunction`....)

**Comment by chadwain** [comment]
Since the goal of this PR is really to fix a regression in `std.fmt`, I added some new tests to `std.fmt` and renamed this PR.

**Pull request merged**